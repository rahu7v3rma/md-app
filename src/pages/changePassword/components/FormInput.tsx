import React, { FunctionComponent, useState } from 'react';

import { EyeIcon, EyeOpenIcon } from '@/assets/svgs';
import { Input } from '@/shared';

import { changePasswordStyles } from '../style';

type Props = {
    placeholder?: string;
    errorMessage?: string;
    showError?: boolean;
    onChangeText?: (text: string) => void;
    onBlur?: () => void;
};

const FormInput: FunctionComponent<Props> = ({
    placeholder,
    errorMessage,
    showError = false,
    onChangeText,
    onBlur
}: Props) => {
    const [secureTextEntry, setTextSecureEntry] = useState(true);

    const handleTextSecureEntry = () => {
        setTextSecureEntry(!secureTextEntry);
    };

    return (
        <Input
            placeholder={placeholder}
            style={changePasswordStyles.input}
            secureTextEntry={secureTextEntry}
            onChangeText={onChangeText}
            Icon={secureTextEntry ? EyeIcon : EyeOpenIcon}
            errorMessage={errorMessage}
            required
            showError={showError}
            iconPress={handleTextSecureEntry}
            onBlur={onBlur}
            textInputStyle={changePasswordStyles.textInput}
        />
    );
};

export default FormInput;
