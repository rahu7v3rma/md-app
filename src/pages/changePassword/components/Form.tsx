import React, { FunctionComponent, useCallback, useState } from 'react';
import { View } from 'react-native';

import { changePasswordStyles } from '../style';

import FormInput from './FormInput';

interface IPasswordState {
    oldPass: string;
    newPass: string;
    confirmPass: string;
}

type Props = {
    onUpdatePasswordState: (passwordState: IPasswordState) => void;
};

const Form: FunctionComponent<Props> = ({ onUpdatePasswordState }: Props) => {
    const [passwordState, setPasswordState] = useState<IPasswordState>({
        oldPass: '',
        newPass: '',
        confirmPass: ''
    });
    const [verifyPassError, setVerifiedPassError] = useState('');

    const updatePassword = useCallback(
        (value: string, key: string) => {
            setPasswordState({
                ...passwordState,
                [key]: value
            });

            // validate that new and confirm passes match if one of them was changed
            if (['newPass', 'confirmPass'].indexOf(key) !== -1) {
                if (
                    (key === 'newPass' &&
                        value !== passwordState.confirmPass) ||
                    (key === 'confirmPass' && value !== passwordState.newPass)
                ) {
                    if (
                        key === 'newPass' &&
                        value !== passwordState.confirmPass &&
                        passwordState.confirmPass !== ''
                    ) {
                        setVerifiedPassError('Password does not match');
                    }
                } else {
                    setVerifiedPassError('');
                }
            }
            onUpdatePasswordState({
                ...passwordState,
                [key]: value
            });
        },
        [passwordState, onUpdatePasswordState]
    );

    const checkValidation = useCallback(
        (value: string, key: string) => {
            if (['newPass', 'confirmPass'].indexOf(key) !== -1) {
                if (
                    (key === 'newPass' &&
                        value !== passwordState.confirmPass) ||
                    (key === 'confirmPass' && value !== passwordState.newPass)
                ) {
                    setVerifiedPassError('Password does not match');
                } else {
                    setVerifiedPassError('');
                }
            }
        },
        [passwordState]
    );

    return (
        <View style={changePasswordStyles.inputContainer}>
            <FormInput
                placeholder="Enter old password*"
                errorMessage="*Required"
                onChangeText={(txt) => updatePassword(txt, 'oldPass')}
            />
            <FormInput
                placeholder="Set up new password*"
                errorMessage={
                    passwordState.newPass === '' ? '*Required' : verifyPassError
                }
                showError={verifyPassError !== ''}
                onChangeText={(txt) => updatePassword(txt, 'newPass')}
            />
            <FormInput
                placeholder="Repeat password*"
                errorMessage={
                    passwordState.confirmPass === ''
                        ? '*Required'
                        : verifyPassError
                }
                onBlur={() =>
                    checkValidation(passwordState.confirmPass, 'confirmPass')
                }
                onChangeText={(txt) => updatePassword(txt, 'confirmPass')}
                showError={verifyPassError !== ''}
            />
        </View>
    );
};

export default Form;
