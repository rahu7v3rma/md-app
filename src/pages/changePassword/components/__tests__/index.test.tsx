import React from 'react';
import renderer, { act } from 'react-test-renderer';

import Form from '../Form';
import FormInput from '../FormInput';

jest.mock('@/assets/svgs', () => {
    const EyeIcon = () => <svg>EyeIcon</svg>;
    const EyeOpenIcon = () => <svg>EyeOpenIcon</svg>;
    return { EyeIcon, EyeOpenIcon };
});

describe('FormInput Component', () => {
    it('renders correctly with a placeholder', () => {
        const component = renderer.create(
            <FormInput placeholder="Test Placeholder" />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('toggles input visibility on icon press', () => {
        const component = renderer.create(
            <FormInput placeholder="Test Placeholder" />
        );
        let tree = component.toJSON();

        const input = tree.children.find((node) => node.type === 'Input');

        if (input) {
            const icon = input.children.find((node) => node.type === 'svg');

            if (icon) {
                const iconPress = icon.props.iconPress;
                iconPress();

                tree = component.toJSON();
                expect(tree).toMatchSnapshot();

                iconPress();
                tree = component.toJSON();
                expect(tree).toMatchSnapshot();
            }
        }
    });

    it('calls the onChangeText function when input changes', () => {
        const onChangeText = jest.fn();
        const component = renderer.create(
            <FormInput
                placeholder="Test Placeholder"
                onChangeText={onChangeText}
            />
        );
        let tree = component.toJSON();

        const input = tree.children.find((node) => node.type === 'Input');

        if (input) {
            const onChangeTextFunction = input.props.onChangeText;

            if (typeof onChangeTextFunction === 'function') {
                onChangeTextFunction('Test Text');
                expect(onChangeText).toHaveBeenCalledWith('Test Text');
            }
        }
    });

    it('calls the onBlur function when input loses focus', () => {
        const onBlur = jest.fn();
        const component = renderer.create(
            <FormInput placeholder="Test Placeholder" onBlur={onBlur} />
        );
        let tree = component.toJSON();

        // Find the input element
        const input = tree.children.find((node) => node.type === 'Input');

        if (input) {
            const onBlurFunction = input.props.onBlur;

            if (typeof onBlurFunction === 'function') {
                onBlurFunction();
                expect(onBlur).toHaveBeenCalled();
            }
        }
    });

    it('displays an error message when showError is true', () => {
        const component = renderer.create(
            <FormInput
                placeholder="Test Placeholder"
                showError={true}
                errorMessage="Test Error"
            />
        );
        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
    });
});

describe('Form Component', () => {
    it('renders correctly with all input fields', async () => {
        const tree = renderer.create(<Form onUpdatePasswordState={() => {}} />);
        expect(tree.root).toBeTruthy();
    });

    it('updates password state when input fields change', async () => {
        const onUpdatePasswordState = jest.fn();
        const tree = renderer.create(
            <Form onUpdatePasswordState={onUpdatePasswordState} />
        );
        const textInput = tree.root.findByProps({
            placeholder: 'Enter old password*'
        }).props;
        await act(() => textInput.onChangeText('pass'));
        expect(onUpdatePasswordState).toHaveBeenCalledWith({
            confirmPass: '',
            newPass: '',
            oldPass: 'pass'
        });
        const newPass = tree.root.findByProps({
            placeholder: 'Set up new password*'
        }).props;
        await act(() => newPass.onChangeText('new pass'));
        expect(onUpdatePasswordState).toHaveBeenCalledWith({
            confirmPass: '',
            newPass: 'new pass',
            oldPass: 'pass'
        });
        const newCnfPass = tree.root.findByProps({
            placeholder: 'Repeat password*'
        }).props;
        await act(() => newCnfPass.onChangeText('new cnf pass'));
        expect(onUpdatePasswordState).toHaveBeenCalledWith({
            confirmPass: 'new cnf pass',
            newPass: 'new pass',
            oldPass: 'pass'
        });
    });

    it('displays error message when new password and confirm password do not match', async () => {
        const onUpdatePasswordState = jest.fn();
        const tree = renderer.create(
            <Form onUpdatePasswordState={onUpdatePasswordState} />
        );
        const textInput = tree.root.findByProps({
            placeholder: 'Enter old password*'
        }).props;
        await act(() => textInput.onChangeText('pass'));
        expect(onUpdatePasswordState).toHaveBeenCalledWith({
            confirmPass: '',
            newPass: '',
            oldPass: 'pass'
        });
        const newPass = tree.root.findByProps({
            placeholder: 'Set up new password*'
        }).props;
        await act(() => newPass.onChangeText('new pass'));
        expect(onUpdatePasswordState).toHaveBeenCalledWith({
            confirmPass: '',
            newPass: 'new pass',
            oldPass: 'pass'
        });
        const newCnfPass = tree.root.findByProps({
            placeholder: 'Repeat password*'
        }).props;
        await act(() => newCnfPass.onChangeText('new cnf pass'));
        expect(onUpdatePasswordState).toHaveBeenCalledWith({
            confirmPass: 'new cnf pass',
            newPass: 'new pass',
            oldPass: 'pass'
        });
        await act(() => newCnfPass.onBlur());
        const newCnfPassUpdate = tree.root.findByProps({
            placeholder: 'Repeat password*'
        }).props;
        expect(newCnfPassUpdate.showError).toBe(true);
        expect(newCnfPassUpdate.errorMessage).toBe('Password does not match');
    });

    it('clears error message when new password and confirm password match', async () => {
        const onUpdatePasswordState = jest.fn();
        const tree = renderer.create(
            <Form onUpdatePasswordState={onUpdatePasswordState} />
        );
        const textInput = tree.root.findByProps({
            placeholder: 'Enter old password*'
        }).props;
        await act(() => textInput.onChangeText('pass'));
        expect(onUpdatePasswordState).toHaveBeenCalledWith({
            confirmPass: '',
            newPass: '',
            oldPass: 'pass'
        });
        const newPass = tree.root.findByProps({
            placeholder: 'Set up new password*'
        }).props;
        await act(() => newPass.onChangeText('new pass'));
        expect(onUpdatePasswordState).toHaveBeenCalledWith({
            confirmPass: '',
            newPass: 'new pass',
            oldPass: 'pass'
        });
        const newCnfPass = tree.root.findByProps({
            placeholder: 'Repeat password*'
        }).props;
        await act(() => newCnfPass.onChangeText('new pass'));
        expect(onUpdatePasswordState).toHaveBeenCalledWith({
            confirmPass: 'new pass',
            newPass: 'new pass',
            oldPass: 'pass'
        });
        const newCnfPassUpdate = tree.root.findByProps({
            placeholder: 'Repeat password*'
        }).props;
        expect(newCnfPassUpdate.showError).toBe(false);
        expect(newCnfPassUpdate.errorMessage).toBe('');
    });

    it('calls the onBlur function when confirming password field loses focus', async () => {
        const onUpdatePasswordState = jest.fn();
        const tree = renderer.create(
            <Form onUpdatePasswordState={onUpdatePasswordState} />
        );
        const newPass = tree.root.findByProps({
            placeholder: 'Set up new password*'
        }).props;
        await act(() => newPass.onChangeText('new pass'));
        expect(onUpdatePasswordState).toHaveBeenCalledWith({
            confirmPass: '',
            newPass: 'new pass',
            oldPass: ''
        });
        const newCnfPass = tree.root.findByProps({
            placeholder: 'Repeat password*'
        }).props;
        await act(() => newCnfPass.onChangeText('new pass'));
        expect(onUpdatePasswordState).toHaveBeenCalledWith({
            confirmPass: 'new pass',
            newPass: 'new pass',
            oldPass: ''
        });
        expect(newCnfPass.showError).toBe(false);

        // on blur we are checking validation so put assertation of error before it
        await act(() => newCnfPass.onBlur());
        const newCnfPassUpdate = tree.root.findByProps({
            placeholder: 'Repeat password*'
        }).props;
        expect(newCnfPassUpdate.showError).toBe(true);
        expect(newCnfPassUpdate.errorMessage).toBe('Password does not match');
    });
});
