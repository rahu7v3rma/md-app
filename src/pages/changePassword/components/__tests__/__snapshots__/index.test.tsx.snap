// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput Component displays an error message when showError is true 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "borderColor": "transparent",
        "borderRadius": 20,
        "borderWidth": 2,
        "elevation": 2,
        "flexDirection": "row",
        "height": 61,
        "justifyContent": "space-between",
        "paddingHorizontal": 10,
        "shadowColor": "rgba(0, 0, 0, 0.3)",
        "shadowOffset": {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0.5,
        "shadowRadius": 1,
      },
      {
        "backgroundColor": "#FFECEC",
        "borderColor": "#F9D4D4",
      },
      {
        "borderColor": "#72996B",
      },
      {
        "marginVertical": 5,
      },
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  />
  <View
    style={
      [
        {
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": "center",
          "paddingHorizontal": 10,
          "position": "relative",
        },
      ]
    }
  >
    <TextInput
      autoCapitalize="none"
      autoCorrect={true}
      autoFocus={false}
      editable={true}
      keyboardType="default"
      multiline={false}
      numberOfLines={1}
      onBlur={[Function]}
      onChange={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Test Placeholder"
      placeholderTextColor="#D05151"
      returnKeyType="done"
      secureTextEntry={true}
      selectionColor="#72996B"
      style={
        [
          {
            "color": "#4D4F4D",
            "flex": 1,
            "fontFamily": "Poppins",
            "fontSize": 16,
            "fontWeight": "500",
          },
          {
            "paddingBottom": 20,
          },
          {
            "fontFamily": "Poppins",
            "fontSize": 16,
            "fontWeight": "500",
            "lineHeight": 20,
          },
        ]
      }
      value=""
    />
    <Text
      adjustsFontSizeToFit={true}
      style={
        [
          {
            "fontFamily": "Poppins",
          },
          {
            "bottom": 12,
            "left": 8,
            "marginBottom": 5,
            "position": "absolute",
            "zIndex": -1,
          },
          {
            "color": "#D05151",
            "fontSize": 12,
            "fontWeight": "500",
            "lineHeight": 20,
          },
        ]
      }
    >
      Test Error
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "marginRight": 5,
        "opacity": 1,
      }
    }
  >
    <svg>
      EyeIcon
    </svg>
  </View>
</View>
`;

exports[`FormInput Component renders correctly with a placeholder 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "borderColor": "transparent",
        "borderRadius": 20,
        "borderWidth": 2,
        "elevation": 2,
        "flexDirection": "row",
        "height": 61,
        "justifyContent": "space-between",
        "paddingHorizontal": 10,
        "shadowColor": "rgba(0, 0, 0, 0.3)",
        "shadowOffset": {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0.5,
        "shadowRadius": 1,
      },
      {},
      false,
      {
        "marginVertical": 5,
      },
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  />
  <View
    style={
      [
        {
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": "center",
          "paddingHorizontal": 10,
          "position": "relative",
        },
      ]
    }
  >
    <TextInput
      autoCapitalize="none"
      autoCorrect={true}
      autoFocus={false}
      editable={true}
      keyboardType="default"
      multiline={false}
      numberOfLines={1}
      onBlur={[Function]}
      onChange={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Test Placeholder"
      placeholderTextColor="#4D4F4D"
      returnKeyType="done"
      secureTextEntry={true}
      selectionColor="#72996B"
      style={
        [
          {
            "color": "#4D4F4D",
            "flex": 1,
            "fontFamily": "Poppins",
            "fontSize": 16,
            "fontWeight": "500",
          },
          {},
          {
            "fontFamily": "Poppins",
            "fontSize": 16,
            "fontWeight": "500",
            "lineHeight": 20,
          },
        ]
      }
      value=""
    />
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "marginRight": 5,
        "opacity": 1,
      }
    }
  >
    <svg>
      EyeIcon
    </svg>
  </View>
</View>
`;
